### ISSUE TITLE
[TT-1666] Requests to listen_paths without trailing slash will match closest listen_path

### ISSUE THREAD

**Issue by zcmack** [TykTechnologies/tyk#2992]
https://github.com/TykTechnologies/tyk/blob/eb495bfa63fe7516abf238a20ede78b49eb83db0/gateway/api_loader.go#L461

We have many api definitions with similar listen paths, as we tend to namespace as part of the listen path string. We have noticed the behavior that if a URL is mistyped, a route match will be triggered to nearest neighbor. For instance if I request `/namespace/apple` but no definition is present, it will match `/namespace/app` and route the request to that upstream target. The linked code snippet seems intended to be able to match any subpath, such as `/namespace/app/mysubpath` but the trailing slash is not included. 

Knowing that different implementations may choose to include trailing slashes or not in their listen_paths, should this be a configurable option that will amend the matching regex to specify the slash?

**Comment by zcmack** [comment]
I haven't run the test suite against it yet, but changing that line to:
`chainDef.ListenOn = spec.Proxy.ListenPath + "{rest:$|\\/.*}"`

will correct the behavior to only match to a specific api definition and be "less fuzzy." i will gladly open a PR if this is confirmed as an option or a bug.

**Comment by zcmack** [comment]
the more i think about this, the more i feel like it is a bug. i did notice that the majority if examples in the documentation have trailing slashes on the listen_path of the api definition. if the above change is not appropriate, i think amending a trailing slash to the listen path at creation time would also ensure consistency.

**Comment by tedbjorklund** [comment]
To add a little more context to this, our proxy definitions are defined as: ` "listen_path": "/api/v1/third"` (with no trailing slash at the end). Without a fix like proposed above, this allows the following request to result in a 200, when it should be a 404. 

Expected result:
```
➜  ~ curl -k https://tyk.api-management/api/v1/thirdapi/v1/foo -H "apikey: thirdkey"   
404 Not Found
```
```
➜  ~ curl -k https://tyk.api-management/api/v1/third/api/v1/foo -H "apikey: thirdkey"   
{"method":"GET","response":"third-stable"}
```
Actual result:
```
➜  ~ curl -k https://tyk.api-management/api/v1/thirdapi/v1/foo -H "apikey: thirdkey"   
{"method":"GET","response":"third-stable"}
```
```
➜  ~ curl -k https://tyk.api-management/api/v1/third/api/v1/foo -H "apikey: thirdkey"   
{"method":"GET","response":"third-stable"}
```

Here is the full proxy definition: 
```
{
    "CORS": {
        "debug": false,
        "enable": false,
        "max_age": 0,
        "allowed_headers": null,
        "allowed_methods": null,
        "allowed_origins": null,
        "exposed_headers": null,
        "allow_credentials": false,
        "options_passthrough": false
    },
    "auth": {
        "signature": {
            "header": "",
            "secret": "",
            "algorithm": "",
            "error_code": 0,
            "error_message": "",
            "allowed_clock_skew": 0
        },
        "use_param": false,
        "param_name": "",
        "use_cookie": false,
        "cookie_name": "",
        "use_certificate": false,
        "auth_header_name": "apikey",
        "validate_signature": false
    },
    "name": "third-api-proxy",
    "slug": "",
    "proxy": {
        "transport": {
            "proxy_url": "",
            "ssl_ciphers": null,
            "ssl_min_version": 0,
            "ssl_insecure_skip_verify": false
        },
        "target_url": "",
        "listen_path": "/api/v1/third",
        "target_list": null,
        "service_discovery": {
            "data_path": "",
            "target_path": "",
            "cache_timeout": 0,
            "port_data_path": "",
            "query_endpoint": "",
            "use_target_list": false,
            "parent_data_path": "",
            "use_nested_query": false,
            "endpoint_returns_list": false,
            "use_discovery_service": false
        },
        "strip_listen_path": true,
        "disable_strip_slash": false,
        "preserve_host_header": false,
        "enable_load_balancing": false,
        "check_host_against_uptime_tests": false
    },
    "active": true,
    "api_id": "third-api-proxy.sample-third",
    "domain": "example.com",
    "org_id": "",
    "internal": false,
    "protocol": "",
    "basic_auth": {
        "cache_ttl": 0,
        "disable_caching": false,
        "body_user_regexp": "",
        "extract_from_body": false,
        "body_password_regexp": ""
    },
    "definition": {
        "key": "x-nm-deployment",
        "location": "header",
        "strip_path": false
    },
    "enable_jwt": false,
    "jwt_source": "",
    "oauth_meta": {
        "auth_login_redirect": "",
        "allowed_access_types": null,
        "allowed_authorize_types": null
    },
    "use_oauth2": false,
    "use_openid": false,
    "allowed_ips": null,
    "config_data": null,
    "listen_port": 0,
    "tag_headers": null,
    "use_keyless": false,
    "certificates": null,
    "do_not_track": false,
    "jwt_skip_kid": false,
    "uptime_tests": {
        "config": {
            "recheck_wait": 0,
            "service_discovery": {
                "data_path": "",
                "target_path": "",
                "cache_timeout": 0,
                "port_data_path": "",
                "query_endpoint": "",
                "use_target_list": false,
                "parent_data_path": "",
                "use_nested_query": false,
                "endpoint_returns_list": false,
                "use_discovery_service": false
            },
            "expire_utime_after": 0
        },
        "check_list": null
    },
    "version_data": {
        "versions": {
            "stable": {
                "name": "stable",
                "paths": {
                    "ignored": null,
                    "black_list": null,
                    "white_list": null
                },
                "expires": "3000-01-02 15:04",
                "extended_paths": {},
                "global_headers": null,
                "override_target": "https://third-api.sample-third.svc.cluster.local/",
                "global_size_limit": 0,
                "use_extended_paths": true,
                "global_headers_remove": null
            },
            "pre-release": {
                "name": "pre-release",
                "paths": {
                    "ignored": null,
                    "black_list": null,
                    "white_list": null
                },
                "expires": "3000-01-02 15:04",
                "extended_paths": {},
                "global_headers": null,
                "override_target": "https://prethird-api.sample-third.svc.cluster.local/",
                "global_size_limit": 0,
                "use_extended_paths": true,
                "global_headers_remove": null
            }
        },
        "not_versioned": false,
        "default_version": "stable"
    },
    "auth_provider": {
        "meta": null,
        "name": "",
        "storage_engine": ""
    },
    "cache_options": {
        "enable_cache": false,
        "cache_timeout": 0,
        "cache_response_codes": null,
        "cache_all_safe_requests": false,
        "cache_control_ttl_header": "",
        "enable_upstream_cache_control": false
    },
    "disable_quota": true,
    "notifications": {
        "shared_secret": "",
        "oauth_on_keychange_url": ""
    },
    "event_handlers": {
        "events": null
    },
    "openid_options": {
        "providers": null,
        "segregate_by_client": false
    },
    "use_basic_auth": false,
    "blacklisted_ips": null,
    "request_signing": {
        "key_id": "",
        "secret": "",
        "algorithm": "",
        "is_enabled": false
    },
    "strip_auth_data": false,
    "session_lifetime": 0,
    "session_provider": {
        "meta": null,
        "name": "",
        "storage_engine": ""
    },
    "custom_middleware": {
        "pre": null,
        "post": null,
        "driver": "",
        "response": null,
        "auth_check": {
            "name": "",
            "path": "",
            "raw_body_only": false,
            "require_session": false
        },
        "id_extractor": {
            "extract_from": "",
            "extract_with": "",
            "extractor_config": null
        },
        "post_key_auth": null
    },
    "global_rate_limit": {
        "per": 0,
        "rate": 0
    },
    "use_standard_auth": false,
    "disable_rate_limit": true,
    "jwt_signing_method": "",
    "pinned_public_keys": null,
    "use_go_plugin_auth": false,
    "client_certificates": null,
    "enable_context_vars": false,
    "response_processors": null,
    "use_mutual_tls_auth": false,
    "jwt_default_policies": null,
    "jwt_scope_claim_name": "",
    "enable_coprocess_auth": false,
    "enable_proxy_protocol": false,
    "jwt_client_base_field": "",
    "jwt_policy_field_name": "",
    "upstream_certificates": {
        "*.sample-third.svc.cluster.local": "/var/run/mounts/kubernetes.io/sample-third/sample-third-lastmile.pem"
    },
    "enable_ip_blacklisting": false,
    "enable_ip_whitelisting": false,
    "expire_analytics_after": 0,
    "hmac_allowed_algorithms": null,
    "hmac_allowed_clock_skew": 0,
    "jwt_identity_base_field": "",
    "custom_middleware_bundle": "",
    "dont_set_quota_on_create": false,
    "base_identity_provided_by": "",
    "enable_signature_checking": false,
    "jwt_scope_to_policy_mapping": null,
    "enable_batch_request_support": false,
    "jwt_issued_at_validation_skew": 0,
    "jwt_expires_at_validation_skew": 0,
    "jwt_not_before_validation_skew": 0
}
```

**Comment by andyo-tyk** [comment]
This was fixed in Tyk 4.0.4.

I'm not sure why this ticket wasn't updated, apologies for the delay.
